packages/ap-ui/alliance_platform/ui/templatetags/alliance_platform/ui.py

```diff
@@ -65,15 +73,76 @@ def add_arg(self, arg: Any):
     def add_kwargs(self, kwargs: dict[str, Any]):
         self.kwargs.update(kwargs)
 
+    def add_params(self, params: QueryParams):
+        if not params:
+            return
+
+        if isinstance(params, str):
+            query_string = params
+        elif isinstance(params, QueryDict):
+            query_string = params.urlencode()
+        else:
+            query_string = urlencode(params)
+
+        # query string will be blank if no params have been added
+        if not self.query_string:
+            self.query_string = "?"
+        elif query_string and not query_string.startswith("&") and not self.query_string.endswith("&"):
+            self.query_string += "&"
+
+        self.query_string += query_string
+
     def resolve(self, context: Context):
         if self.check_perm:
             url = reverse_if_probably_allowed(
-                context["request"], self.url_name, object=self.object, args=self.args, kwargs=self.kwargs
+                context["request"],
+                self.url_name,
+                object=self.object,
+                args=self.args,
+                kwargs=self.kwargs,
             )
             if url is None:
                 raise OmitComponentFromRendering()
-            return url
-        return reverse(self.url_name, args=self.args, kwargs=self.kwargs)
+        else:
+            url = reverse(self.url_name, args=self.args, kwargs=self.kwargs)
+
+        if self.query_string:
+            url += self.query_string
+
+        return url
+
+
+class QueryParamsNode(template.Node):
+    def __init__(
+        self,
+        params: dict[str, Any],
+        target_var=None,
+    ):
+        if not target_var:
+            raise template.TemplateSyntaxError("Specify variable name for query params to be passed as")
+        self.params = params
+        self.target_var = target_var
+
+    def render(self, context):
+        params = {key: value.resolve(context) for key, value in self.params.items()}
+        context[self.target_var] = params
+        return ""
+
+
+@register.tag("query_params")
+def query_params(parser: template.base.Parser, token: template.base.Token):
+    """Pass arbitrary keyword arguments to generate a dictionary with the corresponding keys.
+
+    Needs to be assigned a target var with ``{% query_params as variable %}`` to be
+    passed to a url.
+    """
+    _, kwargs, target_var = parse_tag_arguments(
+        parser,
+        token,
+        supports_as=True,
+    )
+
+    return QueryParamsNode(params=kwargs, target_var=target_var)
```

Comment:
hmm so nothing here looks specific to query - it's really a tag to create a dict? maybe we should make it generic, could be useful for other things where you want to generate a dict in a template to pass through, eg. pass props to a component tag